### PR TITLE
Add support for various NAT traversal schemes

### PR DESCRIPTION
**Pull request by Vagabond**
This PR adds the following NAT traversal schemes:

* NAT uPNP support #24 
* NAT-PMP support
* TCP hole punching using reuseaddr/reuseport #26 

Additionally, it uses something like STUN to determine and validate the external IP/port is dialable externally. (note that this last piece doesn't work right now because we need to originate the connection test from a new local port or possibly a new IP).

### CODE REVIEW

**Comment by Vagabond** [comment]
OK, this is now rebased and uses the `{unique, true}` option in an appropriate place. It still needs more work:

* 2 unit tests fail
* dialyzer warning about `libp2p_identify:observed_addr` taking an opaque term
* distinguish the various flavors of NAT from each other

Beyond those, we need a TURN/relay design and also a way to do 'mutual dialing' where both sides use the STUN derived external IP/port combination to mutually dial each other to punch a hole in restricted cone NAT.

File: src/libp2p_transport_tcp.erl [hunk 1]
```diff
@@ -39,10 +39,11 @@ start_listener(Sup, Addr, TID) ->
             Options = libp2p_config:get_config(?CONFIG_SECTION, OptionDefaults),
             {TransportOpts, ListenOpts0} =
                 lists:partition(fun({Key, _}) ->
-                                        sets:is_element(Key, TransportKeys)
+                                        sets:is_element(Key, TransportKeys);
+                                   (_) -> true
```
Comment (madninja) [inline src/libp2p_transport_tcp.erl]: Can this be removed? What other options are you throwing in there?

File: src/libp2p_transport_tcp.erl [hunk 2]
```diff
@@ -67,16 +68,34 @@ new_connection(Socket) ->
     libp2p_connection:new(?MODULE, #tcp_state{socket=Socket, transport=ranch_tcp}).
 
 -spec dial(string(), [libp2p_swarm:connect_opt()], pos_integer()) -> {ok, libp2p_connection:connection()} | {error, term()}.
-dial(MAddr, _DialOptions, Timeout) ->
+dial(MAddr, DialOptions, Timeout) ->
     case tcp_addr(MAddr) of
         {IP, Port, Type} ->
-            case ranch_tcp:connect(IP, Port, [Type], Timeout) of
+            UniquePort = proplists:get_value(unique_port, DialOptions, false),
```
Comment (madninja) [inline src/libp2p_transport_tcp.erl]: Why not just use `unique` here instead of introducing `unique_port`?

File: src/libp2p_transport_tcp.erl [hunk 3]
```diff
@@ -67,16 +68,34 @@ new_connection(Socket) ->
...
+            Options = case Type == inet of
```
Comment (madninja) [inline src/libp2p_transport_tcp.erl]: I'm not sure this is right? Type can be `inet` or `inet6`

File: src/libp2p_transport_tcp.erl [hunk 4]
```diff
@@ -85,13 +104,18 @@ tcp_listen_addrs(Socket) ->
...
-            [multiaddr({Addr, Port}) ||
+            ListenAddrs = [multiaddr({Addr, Port}) ||
                 {_, Opts} <- IFAddrs, {addr, Addr} <- Opts, {flags, Flags} <- Opts,
                 size(Addr) == size(IP),
                 not lists:member(loopback, Flags),
                 %% filter out ipv6 link-local addresses
                 not (size(Addr) == 8 andalso element(1, Addr) == 16#fe80)
-            ]
+            ],
+            Parent = self(),
```
Comment (madninja) [inline src/libp2p_transport_tcp.erl]: I certainly wasn't thinking of kicking these of here? Perhaps move them to where start_listener happens?

File: src/libp2p_transport_tcp.erl (hunk 1 above)
Comment (Vagabond) [reply src/libp2p_transport_tcp.erl]: The `{raw, ...` option is not a 2-tuple.

File: src/libp2p_transport_tcp.erl (hunk 2 above)
Comment (Vagabond) [reply src/libp2p_transport_tcp.erl]: Unique makes sure we don't reuse an existing session, unique_port makes sure we don't use reuseport. They are different operations that will have different behaviour depending on the type of NAT. Specifically restricted NAT vs port restricted NAT.

Comment (madninja) [reply src/libp2p_transport_tcp.erl]: ah, so you can have a unique session and yet reuse a port?

File: src/libp2p_transport_tcp.erl (hunk 1 above)
Comment (madninja) [reply src/libp2p_transport_tcp.erl]: Right the raw option is added below this partition not before?

File: src/libp2p_transport_tcp.erl (hunk 2 above)
Comment (Vagabond) [reply src/libp2p_transport_tcp.erl]: Yes, which would result in the same IP/port combo on both sides, but a *new* socket being opened. This is sort of the magic of reuseport.

**Comment by Vagabond** [comment]
This PR is pretty clean now, the dialyzer failure I believe has to do with the specs around the identify shortcut you added @madninja 

I think we should get this landed and then we can work on the peerbook stuff on top of it.

**Comment by madninja** [comment]
I know it may be a lot but perhaps we can get better coverage for the new bits?

**Comment by Vagabond** [comment]
Open to suggestions on how. I'm not sure how to do it in a remotely portable way...

**Comment by madninja** [comment]
NOTE: I'm going to squash and merge this, to get us back on a common baseline but we need to:

1. Don't have tcp/udp logic in swarm server. The work to plug transports should address this
2. Use peerbook instead of swarm_server to update and maintain peer information

**Pull request merged**